ARCHITECTURE.md

```diff
@@ -112,37 +120,113 @@ during lookup.
+
+NOTE: We did support nested configurations for sub-stores but removed this
+because the maintenance cost did not justify the benefits of this feature.
 
 ### `internal/cui`
 
-TODO: history, issues, outlook
+The name `cui` is an abbreviation for `console-user-interface` and contains
+several helper functions to interact with humans over a text based interface.
 
-### `internal/editor`
+Most of these ask the user to select some item from a selection or provide
+some input.
 
-TODO: security
+NOTE: We used to support rich terminal UIs with arrow navigation and such.
+However all existing libraries that were available without CGO were either
+abandoned or buggy on some platforms and we didn't have any capacity to fix
+them. So we had to remove support for this feature.
 
 ### `internal/queue`
 
-TODO: motivation, usage
+The `queue` package implements a FIFO queue that executed strictly sequential
+in the background. This allows for certain operations, like a git push, to be
+taken out of the critical path wrt. user interactions. The queue will be fully
+processed before the process exits.
+
+### `internal/store/root`
 
-### `internal/store`
+The `root store` package implements an internal password store API that (only)
+supports mounting `leaf` stores. It will forward (almost) all operations to
+it's `leaf` stores (cross stores moves being a notable exception) and do the
+necessary manipulation of the affected path components.
 
-TODO: everything
+This package makes `gopass` multi-store capable.
+
+### `internal/store/leaf`
+
+The `leaf store` package implements a password store that is mostly compatible
+with any other password store implementation (while aiming for interoperability,
+not 100% feature parity). The low-level operations like filesystem and / or
+version control and crypto operations are passed to the configured `storage`
+or `crypto` backend.
 
 ### `internal/tree`
 
-TODO: why
+The `tree` package implements a simple tree structure that can be printed
+similar to the output of the Unix tool `tree`. It does support different
+`gopass` specific properties (like mounts or templates) not easily implemented
+with other tree packages.
 
 ### `internal/updater`
 
-TODO: security
+The `updater` package implements a secure and anonymous self updater.
+
+Note: The self updater contacts GitHub. If that is concern one should use
+other sources, e.g. distro packages.
+
+It retrieves the latest stable release from GitHub, fetches it's metadata
+and verifies the signature against the built-in release signing keyring.
+
+It tries to avoid conflicting with any `gopass` binary managed by the OS
+and refuse to update these.
 
-### `pkg/...`
+### `pkg/`
 
-TODO: everything
+The package `pkg/` contains our public API surface, i.e. packages we want or
+have to expose externally. Some packages (e.g. `otp`) are only exposed because
+they are being used by some of our integrations. Others (e.g. `pinentry` or
+`pwgen`) are designed for wider use.
```

Comment:
Do we have plan to split the wider use packages into their own repo at some point?
Wouldn't that be better?

Reply:
it would be cleaner but also more annoying. I'd still consider it an option, but not a promise.